src/types.ts

```diff
@@ -63,3 +84,71 @@ export interface UserPermission<T> {
 export interface UserPermissionsFn<T> {
   (userPermissions: UserPermission<T>[], rolePermissions: T[]): T[];
 }
+
+/**
+ * Casbin enforcer getter method interface
+ *
+ * This method provides the Casbin config
+ * required to initialise a Casbin enforcer
+ */
+export interface CasbinEnforcerConfigGetterFn {
+  (
+    authUser: IAuthUserWithPermissions,
+    resource: string,
+    isCasbinPolicy?: boolean,
+  ): Promise<CasbinConfig>;
+}
+
+/**
+ * Casbin resource value modifier method interface
+ *
+ * This method can help modify the resource value
+ * for integration with casbin, as per business logic
+ */
+export interface CasbinResourceModifierFn {
+  (pathParams: string[]): Promise<string>;
+}
+
+/**
+ * Casbin config object
+ */
+export interface CasbinConfig {
+  model: string;
+  policy?: string | PostgresAdapter;
+  allowedRes?: ResourcePermissionObject[];
+}
+
+export interface ResourcePermissionObject {
+  resource: string;
+  permission: string;
+}
+
+export interface IAuthUser {
+  id?: number | string;
+  username: string;
+  password?: string;
+}
+
+export interface IUserPrefs {
+  locale?: string;
+}
+
+export interface IAuthUserWithPermissions<
+  ID = string,
+  TID = string,
+  UTID = string
+  > extends IAuthUser {
+  id?: string;
+  identifier?: ID;
+  permissions: string[];
+  authClientId: number;
+  userPreferences?: IUserPrefs;
+  email?: string;
+  role: string;
+  firstName: string;
+  lastName: string;
+  middleName?: string;
+  tenantId?: TID;
```

Comment:
can we eliminate tenantid, usertenantid and password expiry from here ? That's too app specific

Reply:
done